cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java

```diff
@@ -194,6 +197,31 @@ public KubeTwillRunnerService(MasterEnvironmentContext masterEnvContext, ApiClie
     this.workloadIdentityProvider = workloadIdentityProvider;
   }
 
+  /**
+   * Returns true if the pod is controlled by a stateful set.
+   */
+  private boolean isPodControlledByStatefulSet() {
+    return this.podInfo.getOwnerReferences().stream().anyMatch(r -> r.getKind().equals("StatefulSet"));
+  }
+
+  /**
+   * Generates k8s label selector which selects all runs started by the k8s twill runner that has the run id label.
+   * If the pod is controlled by a StatefulSet it also adds a PARENT_POD_LABEL to the selector.
+   * The PARENT_POD_LABEL is meant to be an identifier for the resources created by the current pod.
+   * Unlike a Deployment, a StatefulSet maintains a sticky identity for each of their Pods.
+   * (https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+   * Therefore, even if the pod goes down, it will come up with the same pod name and this selector can
+   * be continued to be used.
+   */
+  private String generateSelector() {
+    if (isPodControlledByStatefulSet()) {
+      return String.format("%s=%s,%s=%s,%s", PARENT_POD_LABEL, podInfo.getName(),
+                  RUNNER_LABEL, RUNNER_LABEL_VAL, RUN_ID_LABEL);
+    } else {
+      return String.format("%s=%s,%s", RUNNER_LABEL, RUNNER_LABEL_VAL, RUN_ID_LABEL);
```

Comment:
I don't think we need to allow it at all, just error out

Reply:
Done.